shell/server/permissions.js

```diff
@@ -0,0 +1,212 @@
+  // to share results between calls. It may also make sense to cache some graph-connectivity
+  // information from the`mayOpenGrain()` computation to use here.
+
+  var grain = Grains.findOne(grainId);
+  var owner = grain.userId;
+
+  // userId -> PermissionSet
+  var permissionsMap = {}
+  permissionsMap[owner] = roleAssignmentPermissions({allAccess: null}, viewInfo);
+
+  var userStack = [owner];
+  var user;
+
+  while (userStack.length > 0) {
+    var user = userStack.pop();
+    RoleAssignments.find({active: true, sharer: user,
+                          grainId : grainId}).forEach(function (outEdge) {
+      var recipient = outEdge.recipient;
+      var sharer = outEdge.sharer;
+      var changed = false;
+      if (!permissionsMap[recipient]) {
+        changed = true;
+        permissionsMap[recipient] = new PermissionSet();
+      }
+      var newPermissions = roleAssignmentPermissions(outEdge.roleAssignment, viewInfo);
+      newPermissions.intersect(permissionsMap[sharer]);
+      if (permissionsMap[recipient].add(newPermissions)) {
+        changed = true;
+      }
+      if (changed) {
+        userStack.push(recipient);
+      }
+    });
+  }
+
+  if (permissionsMap[userId]) {
+    return permissionsMap[userId];
+  } else {
+    return new PermissionSet();
+  }
+}
+
+apiTokenPermissions = function (token, viewInfo) {
+  var result = grainPermissionsInternal(token.grainId, token.userId, viewInfo);
+  var edge = roleAssignmentPermissions(token.roleAssignment, viewInfo);
+  result.intersect(edge);
+  return result.array;
+}
+
+grainPermissions = function (grainId, userId, viewInfo) {
+  return grainPermissionsInternal(grainId, userId, viewInfo).array;
+}
+
+mayOpenGrain = function(grainId, userId) {
+  // Determines whether the user is allowed to open the grain by searching depth first
+  // for a path of active role assignments leading from the grain owner to the user.
+
+  var grain = Grains.findOne(grainId);
+  if (!grain) { return false; }
+  if (!grain.private) { return true; }
+  if (!userId) { return false; }
+  var owner = grain.userId;
+  if (userId == owner) {
+    return true;
+  }
+
+  var stackedUsers = {userId : true};
+  var userStack = [userId];
+
+  while (userStack.length > 0) {
+    var user = userStack.pop();
+    var edges = RoleAssignments.find({active: true, recipient: user, grainId : grainId}).fetch();
+    for (var ii = 0; ii < edges.length; ++ii) {
+      var inEdge = edges[ii];
+      var sharer = inEdge.sharer;
+      if (sharer == owner) {
+        return true;
+      }
+      if (!stackedUsers[sharer]) {
+        userStack.push(sharer);
+        stackedUsers[sharer] = true;
+      }
+    }
+  }
+  return false;
+}
+
+Meteor.startup(function () {
+  RoleAssignments.find().observe({
+    changed : function (newRoleAssignment, oldRoleAssignment) {
+      if (newRoleAssignment.active != oldRoleAssignment.active) {
```

Comment:
What if assigned role is modified but the assignment remains active? Or do you think that should be implemented by removing the old assignment and creating a new one? IMO we should be able to change them in-place.

Also we probably only want to reset the user's session if their overall permissions changed (and maybe only if their permissions changed negatively?) to avoid disruptions when someone shares to you a permission you already had.